### PR TITLE
feat: adding error codes to ShellJS

### PR DESCRIPTION
**Pull request by nfischer**
This is a redo of #269.

This adds error codes as a `.code` attribute to ShellStrings, but doesn't implement the `errorCode()` function.

This also does not switch the implementation of the `error()` function (that should be in another PR, IMO).

Tests will be added soon.

### CODE REVIEW

**Comment by nfischer** [comment]
@ariporad @levithomason This should be ready for review.

This adds the `.code` attribute to everything.

If we like this implementation and merge this, we can open a new PR to change `error()` to return the code instead of the last command's error string. I didn't include it in this PR because error codes themselves shouldn't be a breaking change, and it's easy enough to swap that implementation in a separate PR.

I left this as 2 commits, since the first one adds error codes, and the second one refactors everything to return a ShellString (which gives every command the `.code` and `.stderr` attributes). Feel free to squash if you prefer.

File: test/pipe.js [hunk 1]
```diff
@@ -11,13 +11,6 @@ shell.mkdir('tmp');
 // Invalids
 //
 
-// no piped methods for commands that don't return anything
-assert.throws(function() {
```
Comment (ariporad) [inline test/pipe.js]: Why did this get removed?

Comment (nfischer) [inline test/pipe.js]: Almost every command returns a ShellString now. So the number of "commands that don't return anything" has shrunk to almost zero.

Comment (ariporad) [inline test/pipe.js]: OK.

**Comment by ariporad** [comment]
LGTM!

**Pull request merged**